backend/src/v2/driver/driver.go

```diff
@@ -671,13 +671,26 @@ func resolveInputs(ctx context.Context, dag *metadata.DAG, iterationIndex *int,
 				}
 			}
 			value, hasValue := inputs.GetParameterValues()[name]
-			if !hasValue {
+
+			// Handle when parameter does not have input value
+			if !hasValue && inputsSpec.GetParameters()[name].IsOptional == false {
+				// when parameter is not optional, value comes from default value, otherwise error
 				if spec.GetDefaultValue() == nil {
-					return fmt.Errorf("input parameter %q is required", name)
+					return fmt.Errorf("no value or default value provided for non-optional parameter %q", name)
 				}
 				inputs.GetParameterValues()[name] = spec.GetDefaultValue()
 				value = spec.GetDefaultValue()
+			} else if !hasValue && inputsSpec.GetParameters()[name].IsOptional == true {
+				// when parameter is optional, value comes from default value, otherwise we delete this parameter from input list, because nothing is being passed
```

Comment:
comment isn't accurate, "optional + no value" means values comes from default. 
Can we avoid the deletion part--only added it when needed instead of add first then delete later if applicable?

Reply:
Thank you for catching this, I updated the logic and comments. I also removed the deletion part, since it doesn't seem to exist in the first place.